resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/kubernetes/submit/submitsteps/hadoopsteps/HadoopKerberosKeytabResolverStep.scala

```diff
@@ -0,0 +1,164 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.spark.deploy.kubernetes.submit.submitsteps.hadoopsteps
+
+import java.io._
+import java.security.PrivilegedExceptionAction
+
+import scala.collection.JavaConverters._
+import scala.util.Try
+
+import io.fabric8.kubernetes.api.model.SecretBuilder
+import org.apache.commons.codec.binary.Base64
+import org.apache.hadoop.conf.Configuration
+import org.apache.hadoop.fs.FileSystem
+import org.apache.hadoop.security.{Credentials, UserGroupInformation}
+import org.apache.hadoop.security.token.{Token, TokenIdentifier}
+import org.apache.hadoop.security.token.delegation.AbstractDelegationTokenIdentifier
+
+import org.apache.spark.SparkConf
+import org.apache.spark.deploy.SparkHadoopUtil
+import org.apache.spark.deploy.kubernetes.{KerberosTokenConfBootstrapImpl, PodWithMainContainer}
+import org.apache.spark.deploy.kubernetes.constants._
+import org.apache.spark.internal.Logging
+
+ /**
+  * This step does all the heavy lifting for Delegation Token logic. This step
+  * assumes that the job user has either specified a principal and keytab or ran
+  * $kinit before running spark-submit. With a TGT stored locally, by running
+  * UGI.getCurrentUser you are able to obtain the current user, alternatively
+  * you can run UGI.logingUserFromKeytabAndReturnUGI and by running .doAs run
+  * as the logged into user instead of the current user. With the Job User principal
+  * you then retrieve the delegation token from the NameNode and store values in
+  * DelegationToken. Lastly, the class puts the data into a secret. All this is
+  * appended to the current HadoopSpec which in turn will append to the current
+  * DriverSpec.
+  */
+private[spark] class HadoopKerberosKeytabResolverStep(
+  submissionSparkConf: SparkConf,
+  maybePrincipal: Option[String],
+  maybeKeytab: Option[File]) extends HadoopConfigurationStep with Logging{
+   private var originalCredentials: Credentials = _
+   private var dfs : FileSystem = _
+   private var renewer: String = _
+   private var renewedCredentials: Credentials = _
+   private var renewedTokens: Iterable[Token[_ <: TokenIdentifier]] = _
+   override def configureContainers(hadoopConfigSpec: HadoopConfigSpec): HadoopConfigSpec = {
+    val hadoopConf = SparkHadoopUtil.get.newConfiguration(submissionSparkConf)
+    logInfo(s"Hadoop Configuration: ${hadoopConf.toString}")
+    if (!UserGroupInformation.isSecurityEnabled) logError("Hadoop not configuration with Kerberos")
+    val maybeJobUserUGI =
+      for {
+        principal <- maybePrincipal
+        keytab <- maybeKeytab
+      } yield {
+        // Not necessary with [Spark-16742]
+        // Reliant on [Spark-20328] for changing to YARN principal
+        submissionSparkConf.set("spark.yarn.principal", principal)
```

Comment:
In [Spark-20328] they specified the spark.yarn.principal when calling the `.addDelegationToken()` methods. Whether it was necessary, I will investigate